### PR TITLE
[UAR-486] capture resigned_on and start_date dates for new MOs

### PR DESCRIPTION
**Pull request by corybudd-ch**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-486

### Change description

- Added date inputs to capture the start date and resigned on dates for new Managing Officers added in the update journey. Does not impact adding new MOs in registration journey.

- Also added summary/dropdown text on ceased/resignedOn BOs/MOs.

- Add mapping of start_date and resigned_on information from form data into model.

- Create new ValidationChain specific to Update journey's add new MO routes, as we don't want to validate start_date and resigned_on date in Registration journey.

Screenshot 2023-05-18 at 11 50 03

Screenshot 2023-05-18 at 11 50 45

Screenshot 2023-05-18 at 11 50 54

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/utils/managing.officer.corporate.ts [hunk 1]
```diff
@@ -129,6 +136,15 @@ const setOfficerData = (reqBody: any, id: string): ApplicationDataType => {
     data[PublicRegisterNameKey] = "";
     data[RegistrationNumberKey] = "";
   }
+
+  // only set start_date and resigned_on keys for Update journey
+  if (reqBody['start_date-day']){
+    data[StartDateKey] = mapFieldsToDataObject(reqBody, StartDateKeys, InputDateKeys);
+  }
+  if (reqBody['is_still_mo']){
```
Comment (AndreKramer7654) [inline src/utils/managing.officer.corporate.ts:144]: what if reqBody['is_still_mo'] === 0 - you won't get to line 145. Use contains?

File: src/validation/custom.validation.ts [hunk 2]
```diff
@@ -233,13 +233,14 @@ export const checkMoreThanOneDateFieldIsNotMissing = (dayStr: string = "", month
 
 export const checkCeasedDateOnOrAfterStartDate = (
   ceaseDayStr: string = "", ceaseMonthStr: string = "", ceaseYearStr: string = "",
-  startDayStr: string = "", startMonthStr: string = "", startYearStr: string = ""
+  startDayStr: string = "", startMonthStr: string = "", startYearStr: string = "",
+  errorMessage: string = ""
```
Comment (AndreKramer7654) [inline src/validation/custom.validation.ts:237]: seems dangerous to default error message to "". "date ordering error"?

File: test/__mocks__/fields/date.mock.ts [hunk 3]
```diff
@@ -20,7 +20,13 @@ export const DATE_OF_BIRTH = {
...
     "filing_date-month": "1",
     "filing_date-year": "2022",
-  }, DATE = { day: "1", month: "1", year: "2000" }, EMPTY_DATE = { day: "", month: "", year: "" },
+  }, RESIGNED_ON_DATE = {
+    "resigned_on-day": "1",
+    "resigned_on-month": "1",
+    "resigned_on-year": "2023",
+  },
+
+  DATE = { day: "1", month: "1", year: "2000" }, EMPTY_DATE = { day: "", month: "", year: "" },
```
Comment (AndreKramer7654) [inline test/__mocks__/fields/date.mock.ts:29]: not ours but a newline could move from line 28 to 3rd, on 29

File: src/utils/managing.officer.corporate.ts (hunk 1 above)
Comment (corybudd-ch) [reply src/utils/managing.officer.corporate.ts:144]: will up date to use `'start_date-day' in reqBody` and `'is_still_mo' in reqBody`

File: test/__mocks__/fields/date.mock.ts (hunk 3 above)
Comment (corybudd-ch) [reply test/__mocks__/fields/date.mock.ts]: done!

File: src/utils/managing.officer.individual.ts [hunk 4]
```diff
@@ -128,6 +140,13 @@ const setOfficerData = (reqBody: any, id: string): ApplicationDataType => {
     data[FormerNamesKey] = "";
   }
 
+  // only set start_date and resigned_on keys for Update journey
+  if (reqBody['start_date-day']){
+    data[StartDateKey] = mapFieldsToDataObject(reqBody, StartDateKeys, InputDateKeys);
+  }
+  if (reqBody['is_still_mo']){
```
Comment (AndreKramer7654) [inline src/utils/managing.officer.individual.ts:147]: needs to be "in" too

Comment (corybudd-ch) [reply src/utils/managing.officer.individual.ts]: done

File: src/utils/managing.officer.corporate.ts [hunk 5]
```diff
@@ -16,20 +17,26 @@ import {
...
 export const getManagingOfficerCorporate = (req: Request, res: Response, backLinkUrl: string, templateName: string) => {
   logger.debugRequest(req, `${req.method} ${req.route.path}`);
 
+  const appData: ApplicationData = getApplicationData(req.session as Session);
+
   return res.render(templateName, {
     backLinkUrl: backLinkUrl,
-    templateName: templateName
+    templateName: templateName,
+    entity_number: appData[EntityNumberKey]
```
Comment (stephanie-ch) [inline src/utils/managing.officer.corporate.ts:39]: Any impact on registration here? Looks to be a shared controller

Comment (corybudd-ch) [reply src/utils/managing.officer.corporate.ts:39]: It shouldn't be, it just makes the `entity_number` available to the templates but its just used to drive whether to show the start_date and resigned_on fields or not

File: views/includes/inputs/fields/is-still-a-managing-officer-input.html [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{% if managingOfficerPronoun == "they" %}
+  {% set dateText = "Date they stopped being a managing officer" %}
+  {% set stillMoText = "Are they still a managing officer for the overseas entity?" %}
+{% else %}
+  {% set dateText = "Date it stopped being a managing officer" %}
+  {% set stillMoText = "Is it still a managing officer for the overseas entity?" %}
```
Comment (stephanie-ch) [inline views/includes/inputs/fields/is-still-a-managing-officer-input.html:6]: Can't see a screenshot on ticket. Where you get the content?

Comment (corybudd-ch) [reply views/includes/inputs/fields/is-still-a-managing-officer-input.html:6]: I got them from the prototype links on the ticket but it seems like the password on there no longer works :(

File: src/utils/managing.officer.corporate.ts (hunk 5 above)
Comment (corybudd-ch) [reply src/utils/managing.officer.corporate.ts:39]: And I've ran through the Registration journey to add New MOs and both for individual and corporate its all still as is!

File: views/includes/inputs/fields/is-still-a-managing-officer-input.html (hunk 6 above)
Comment (stephanie-ch) [reply views/includes/inputs/fields/is-still-a-managing-officer-input.html:6]: Okay no bother, I tried to check the prototype and wasn't able to get in but happy with that answer :D

**Review by stephanie-ch — APPROVED** [review]

**Review by olaych — APPROVED** [review]

**Review by AndreKramer7654 — APPROVED** [review]

**Pull request merged**